AP_Scripting: add button binding

IamPete1 opened the pull request:
This adds a binding to assess AP_Button from scripting.

I ended up having to access AP_Button via AP_Vehicle, this is because I couldn't give a AP_Button a singleton because it has multiple instances in copter and plane. I think this is probably a bug @peterbarker?. It is defined in both AP_Vehicle and the vehicle codes, if I remover either of them it works. The difference between rover and plane and copter is that the button params are not in g2 in rover. Using only the AP_Vehicle define breaks the g2 params.

My attempt to fix that is here, but I could not get quite get it working.

https://github.com/IamPete1/ardupilot/commits/button_singlton

WickedShell commented:
What's the use case for this? I'm wondering if you really want the mask, or do you want a method that tells you what the state for a given button is?

Also FWIW the library it's trying to interface doesn't even look like it's doing any debouncing, so you'd have to manage that in scripting anyways, at which point I suspect a generic GPIO read/write abstraction is more useful. But I've never really used AP_Button so I might be missing something obvious about what it does for you.

IamPete1 commented:
rebased on https://github.com/ArduPilot/ardupilot/pull/12912 and done properly with AP_Button singleton. Added get_button_state function to return the state of a given button with crude de-bouncing

IamPete1 commented:
rebased

IamPete1 commented:
rebased

IamPete1 commented:
@Fungraphics This is a build from this branch for Pixhawk1. This is latest/master/dev firmware so is not tested as mush as stable, might be bugs ect.

[ardurover.zip](https://github.com/ArduPilot/ardupilot/files/4105832/ardurover.zip)

IamPete1 commented:
rebased

Fungraphics commented:
Thanks, i ll try to flash it to my Pixhawk 2.46

I managed to compile it from source but i could not make button script work
at the end.Will keep trying?

Anton

On Fri, 28 Feb 2020, 03:11 Peter Hall, <notifications@github.com> wrote:

> rebased
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/12887?email_source=notifications&email_token=ALXVUVRRJQAIU6HNNXK4WYLRFBXGRA5CNFSM4JRJCGH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENGWTAQ#issuecomment-592275842>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALXVUVQHAPKBSLODW737X2DRFBXGRANCNFSM4JRJCGHQ>
> .
>

IamPete1 commented:
@Fungraphics before flashing this version or messing with the script the thing to do is hook up a button and enable it, you should then see the button press reported over mavlink using mission planners mavlink inspector (ctrl + f). Once that is working the we can narrow the issue down to this stuff.

Fungraphics commented:
Thanks a lot. Will try that i found somewhere that some resistor has to be put in line in order to work. 



Od: Peter Hall
Poslano: petek, 28. februar 2020 15:42
Za: ArduPilot/ardupilot
Kp: Fungraphics; Mention
Zadeva: Re: [ArduPilot/ardupilot] AP_Scripting: add button binding (#12887)

@Fungraphics before flashing this version or messing with the script the thing to do is hook up a button and enable it, you should then see the button press reported over mavlink using mission planners mavlink inspector (ctrl + f). Once that is working the we can narrow the issue down to this stuff.
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

IamPete1 commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -205,3 +205,7 @@ singleton AP_Mission method get_prev_nav_cmd_id uint16_t
+
+include AP_Button/AP_Button.h
+singleton AP_Button alias Button
+singleton AP_Button method get_button_state boolean uint8_t 1 4
```
use AP_BUTTON_NUM_PINS

IamPete1 commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -205,3 +205,7 @@ singleton AP_Mission method get_prev_nav_cmd_id uint16_t
 singleton AP_Mission method num_commands uint16_t
+
+include AP_Button/AP_Button.h
+singleton AP_Button alias Button
```
button

IamPete1 commented:
fixed dev-call issues and rebased
